src/Voron/Data/BTrees/Tree.MultiTree.cs

```diff
@@ -30,8 +31,191 @@ namespace Voron.Data.BTrees
+            ValidateValuesForMultiBulkAdd(maxNodeSize, values);
+            
+            // This is a new tree, we've to put a flag in the header.
+            if ((State.Header.Flags & TreeFlags.MultiValueTrees) != TreeFlags.MultiValueTrees)
+            {
+                ref var state = ref State.Modify();
+                state.Flags |= TreeFlags.MultiValueTrees;
+            }
+
+            var page = FindPageFor(key, out _);
+            if (page is not { LastMatch: 0 }) // Key is not in the tree. Use an optimized path for this scenario.
+            {
+                MultiBulkAddOnNewValue(key, values, maxNodeSize);
+                return;
+            }
+
+            ContextBoundNativeList<int> newItems;
+            {
+                var readonlyKeyItem = page.GetNode(page.LastSearchPosition);
+                if (readonlyKeyItem->Flags == TreeNodeFlags.MultiValuePageRef)
+                {
+
+                    // We have an inner tree for the key. Since our values are sorted,
+                    // we can iterate over the elements and add them one by one.
+                    // As the page is cached, it is already loaded into memory. 
+                    // Therefore, this path could be further optimized by batching them into a single page.
+                    var existingTree = OpenMultiValueTree(key, readonlyKeyItem);
+                    for (int i = 0; i < values.Length; i++)
+                    {
+                        existingTree.AddKeyOnly(values[i]).Dispose();
+                    }
+                    return;
+                }
+                
+                if (readonlyKeyItem->Flags == TreeNodeFlags.PageRef)
+                    throw new InvalidOperationException("Multi trees don't use overflows");
+                
+                // We have a nested page here, which means our values are inlined into the page.
+                // Since we want to avoid modifying the page unless it is necessary,
+                // we will prepare a list of new terms based on the loaded data.
+                // This way, we can avoid modifying the page if we don't have to.
+                newItems = GetListOfNewItems(ref readonlyKeyItem, values);
+            }
+
+            if (newItems.Count == 0)
+            {
+                // Everything is already in the tree, nothing to do here.
+                newItems.Dispose();
+                return;
+            }
+
+            // We know there are new insertions. Let's try to insert them.
+            // Since we scoped the readonly keys in the previous scope,
+            // we are guaranteed not to use invalid structures, but we need to recreate them.
+            {
+                page = ModifyPage(page);
+                var keyItem = page.GetNode(page.LastSearchPosition);
+                
+                for (int i = 0; i < newItems.Count; i++)
+                {
+                    GetNestedPagePointer(page, out var nestedPage, out byte* nestedPagePtr);
```

Comment:
Do we need to execute `GetNestedPagePointer` every single time? This call makes sense only when page has changed, and if I am not missing something that only happens on line #171. Can't we just avoid the call altogether unless we hit there?